app/models/physical_server_firmware_update_task/state_machine.rb

```diff
@@ -0,0 +1,35 @@
+module PhysicalServerFirmwareUpdateTask::StateMachine
+  def run_firmware_update
+    dump_obj(options, "MIQ(#{self.class.name}##{__method__}) options: ", $log, :info)
+    signal :start_firmware_update
+  end
+
+  def start_firmware_update
+    # Implement firmware update in subclass, user-defined values are stored in options field.
+    # Affected servers can be accessed via miq_request.affected_physical_servers
+    raise NotImplementedError, 'Must be implemented in subclass and signal :done_firmware_update when done'
+  end
+
+  def done_firmware_update
+    update_and_notify_parent(:message => msg('done updating firmware'))
+    signal :mark_as_completed
+  end
+
+  def mark_as_completed
+    update_and_notify_parent(:state => 'finished', :message => msg('firmware update completed'))
+    MiqEvent.raise_evm_event(self, 'generic_task_finish', :message => "Done updating firmware on PhysicalServer(s)")
```

Comment:
Is it required to raise an event here? Are you going to create policies based on this event?

Reply:
No requirement currently, no, I copy-pasted from provisioning example thinking it's needed for some reason. Let me remove it then

Reply:
Removed